design/aws-ecs-autoscaling-queue-worker.md

```diff
@@ -0,0 +1,225 @@
+# AWS ECS - L3 Construct for Autoscaling ECS/Fargate Service that Processes Items in a SQS Queue
+
+To address issue [#2396](https://github.com/awslabs/aws-cdk/issues/2396), the AWS ECS CDK construct library should provide a way for customers to create a Queue Worker service (an AWS ECS/Fargate Service that processes items from an sqs queue). This construct will create an SQS Queue and enable autoscaling for the service based on cpu usage and the sqs approximateNumberOfMessagesVisible metric. This would mean adding new ECS CDK constructs `Ec2QueueWorkerService` and `FargateQueryWorkerService`, that would take in the necessary properties required to create a Task Definition, an SQS Queue as well as an ECS/Fargate Service with Auto Scaling rules.
+
+## General approach
+
+The new [`ecs.QueueWorkerServiceBase`], [`ecs.Ec2QueueWorkerService`] and [`ecs.FargateQueueWorkerService`] classes will include L3 constructs for:
+
+* Ec2QueueWorkerService
+* FargateQueueWorkerService
+
+A `QueueWorkerService` will create a task definition with the specified container (on both EC2 and Fargate). An `SQSQueue` will be created and autoscaling of the ECS Service will be dependent on both CPU as well as the SQS queue's `ApproximateNumberOfMessagesVisible` metric. 
+
+## Code changes
+
+Given the above, we should make the following changes to support queue workers on ECS (for both EC2 and Fargate):
+1. Create `QueueWorkerServiceBaseProps` interface and `QueueWorkerServiceBase` construct
+2. Create `Ec2QueueWorkerServiceProps` interface  and `Ec2QueueWorkerService` construct
+3. Create `FargateQueueWorkerServiceProps` interface  and `FargateQueueWorkerService` construct
+
+# Part 1: Create `QueueWorkerServiceBaseProps` interface  and `QueueWorkerServiceBase` construct  
```

Comment:
Style nit: `#` in markdown will create a bigger heading than `##`, if this was intended to be a subsection of `## code changes` (When I view the file as markdown, this makes the hierarchy appear confusing).

Reply:
Updated